---
title: "EpiCompare"
date: "<h4>Vignette updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output:
  BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{templateR} 
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir=here::here())
knitr::opts_knit$set(echo = TRUE, root.dir=here::here())

#### CRAN ####
pkgs <- c("dplyr")
for(p in pkgs){
  if(!require(p, character.only = TRUE)) install.packages(p)
} 
#### GitHub ####
if(!require("EpiCompare")) remotes::install_github("neurogenomics/EpiCompare")

source(here::here("code/utils.R"))
root <- "/Volumes/bms20/projects/neurogenomics-lab/live/Data/tip_seq"
```

# Import data 

Import the called peak files produced by the [nf-core/cutanrun](https://nf-co.re/cutandrun) pipeline.

```{r}
peakfiles <- EpiCompare::gather_files(dir = file.path(root,"processed_data"),
                                      type = "peaks.stringent",
                                      nfcore_cutandrun=TRUE,
                                      workers = 10)
names(peakfiles) <- gsub("S_","S",names(peakfiles),ignore.case = TRUE)

picardfiles <- EpiCompare::gather_files(dir = file.path(root,"processed_data"),
                                        type = "picard",
                                        nfcore_cutandrun=TRUE,
                                        workers = 10)
names(picardfiles) <- gsub("S_","S",names(picardfiles),ignore.case = TRUE)
```

### Translate sample names

```{r}  
peakfile_names <- data.frame(peakfile=names(peakfiles)) %>%
    tidyr::separate(col = "peakfile",
                   into = c("batch","sample_replicate"),
                   remove = FALSE, sep = "[.]") %>%
    dplyr::mutate(sample_replicate=gsub("S_","S",toupper(sample_replicate))) %>%
    tidyr::separate(col = "sample_replicate",
                  into = c("sample","replicate"),
                  remove = FALSE, sep = "_") %>%
  tidyr::separate(col = "batch",
                  into = c("V1","phase","day","month","year"),
                  remove = FALSE, sep = "_")  %>%
    dplyr::select(-V1)


sample_meta <- xlsx::read.xlsx(file = file.path(root,"raw_data/bulkTIP-seq/TIPseq_metadata_manual.xlsx"),
                               sheetIndex = 1) %>%
    dplyr::mutate(sample=gsub("S_","S",sample,ignore.case = TRUE)) %>% 
    merge(peakfile_names,  by=c("batch","sample","replicate"), all.x = TRUE) %>%
    dplyr::mutate(label = gsub(" +","",
                                paste(histone.mark,
                                antibody.company, 
                                # cell.line,
                                peakfile,
                                sep="."))
            ) %>%
    data.frame()
sample_dict <- stats::setNames(sample_meta$label, 
                               sample_meta$peakfile)


#### Translate names #### 
names(peakfiles) <- sample_dict[names(peakfiles)] 
## Make sure sorted by histone mark
peakfiles <- peakfiles[sort(names(peakfiles))]
names(picardfiles) <- sample_dict[names(picardfiles)]  
```

### Get ENCODE data

Import the full genome-wide ENCODE data since 
(`encode_H3K27ac` only includes chromosome 1).
- [H3K27ac_K562_ENCODE_ENCFF044JNJ](https://www.encodeproject.org/files/ENCFF044JNJ/) 

Use the function `EpiCompare::import_narrowPeak` 
instead of the usual `rtracklayer::import` because the latter is
unable to parse this particular set of files.

**IMPORTANT**: These ENCODE files are aligned to *hg19*.

```{r} 
#### H3K27ac ####  
encode_ac <- rtracklayer::import("https://www.encodeproject.org/files/ENCFF044JNJ/@@download/ENCFF044JNJ.bed.gz", format = "narrowPeak")

# names(encode1)[1] <- "H3K27ac_K562_ENCODE_ENCFF044JNJ"

#### H3K27me3 ####
encode_me3 <- rtracklayer::import( "https://www.encodeproject.org/files/ENCFF322IFF/@@download/ENCFF322IFF.bed.gz", 
                                  format = "narrowPeak")

ref_list <- list("encode_ac"=encode_ac,
                 "encode_me3"=encode_me3)
```


```{r}
data("hg38_blacklist") # example blacklist  
EpiCompare::EpiCompare(peakfiles = peakfiles, 
                         picard_files = picardfiles,
                         blacklist = hg38_blacklist,
                         genome_build = list(peakfiles="hg38",
                                             reference="hg19",
                                             blacklist="hg38"),
                         genome_build_output = "hg38",
                         reference = ref_list,
                         upset_plot = TRUE,
                         stat_plot = TRUE,
                         chromHMM_plot = TRUE,
                         chromHMM_annotation = "K562",
                         chipseeker_plot = TRUE,
                         enrichment_plot = TRUE,
                         tss_plot = TRUE,
                         save_output = TRUE, 
                         output_dir = here::here("reports"))

```

 
# Session Info 

<details> 

```{r Session Info}
utils::sessionInfo()
```

</details>  

<br>
